crates/storage/provider/src/providers/blockchain_provider.rs

```diff
@@ -0,0 +1,817 @@
+where
+    DB: Database,
+{
+    fn fill_env_at<EvmConfig>(
+        &self,
+        cfg: &mut CfgEnvWithHandlerCfg,
+        block_env: &mut BlockEnv,
+        at: BlockHashOrNumber,
+        evm_config: EvmConfig,
+    ) -> ProviderResult<()>
+    where
+        EvmConfig: ConfigureEvmEnv,
+    {
+        self.database.provider()?.fill_env_at(cfg, block_env, at, evm_config)
+    }
+
+    fn fill_env_with_header<EvmConfig>(
+        &self,
+        cfg: &mut CfgEnvWithHandlerCfg,
+        block_env: &mut BlockEnv,
+        header: &Header,
+        evm_config: EvmConfig,
+    ) -> ProviderResult<()>
+    where
+        EvmConfig: ConfigureEvmEnv,
+    {
+        self.database.provider()?.fill_env_with_header(cfg, block_env, header, evm_config)
+    }
+
+    fn fill_cfg_env_at<EvmConfig>(
+        &self,
+        cfg: &mut CfgEnvWithHandlerCfg,
+        at: BlockHashOrNumber,
+        evm_config: EvmConfig,
+    ) -> ProviderResult<()>
+    where
+        EvmConfig: ConfigureEvmEnv,
+    {
+        self.database.provider()?.fill_cfg_env_at(cfg, at, evm_config)
+    }
+
+    fn fill_cfg_env_with_header<EvmConfig>(
+        &self,
+        cfg: &mut CfgEnvWithHandlerCfg,
+        header: &Header,
+        evm_config: EvmConfig,
+    ) -> ProviderResult<()>
+    where
+        EvmConfig: ConfigureEvmEnv,
+    {
+        self.database.provider()?.fill_cfg_env_with_header(cfg, header, evm_config)
+    }
+}
+
+impl<DB> PruneCheckpointReader for BlockchainProvider2<DB>
+where
+    DB: Database,
+{
+    fn get_prune_checkpoint(
+        &self,
+        segment: PruneSegment,
+    ) -> ProviderResult<Option<PruneCheckpoint>> {
+        self.database.provider()?.get_prune_checkpoint(segment)
+    }
+
+    fn get_prune_checkpoints(&self) -> ProviderResult<Vec<(PruneSegment, PruneCheckpoint)>> {
+        self.database.provider()?.get_prune_checkpoints()
+    }
+}
+
+impl<DB> ChainSpecProvider for BlockchainProvider2<DB>
+where
+    DB: Send + Sync,
+{
+    fn chain_spec(&self) -> Arc<ChainSpec> {
+        self.database.chain_spec()
+    }
+}
+
+impl<DB> StateProviderFactory for BlockchainProvider2<DB>
+where
+    DB: Database,
+{
+    /// Storage provider for latest block
+    fn latest(&self) -> ProviderResult<StateProviderBox> {
+        trace!(target: "providers::blockchain", "Getting latest block state provider");
+        self.database.latest()
+    }
+
+    fn history_by_block_number(
+        &self,
+        block_number: BlockNumber,
+    ) -> ProviderResult<StateProviderBox> {
+        trace!(target: "providers::blockchain", ?block_number, "Getting history by block number");
+        self.ensure_canonical_block(block_number)?;
+        self.database.history_by_block_number(block_number)
+    }
+
+    fn history_by_block_hash(&self, block_hash: BlockHash) -> ProviderResult<StateProviderBox> {
+        trace!(target: "providers::blockchain", ?block_hash, "Getting history by block hash");
+        self.database.history_by_block_hash(block_hash)
+    }
```

Comment:
these would also need to check the in memory provider as well because from caller pov all blocks in there are also historical blocks, part of the canonical chain

Reply:
hmm but these methods are returning a `StateProviderBox` , which is `Box<dyn StateProvider>` and in memory we are only storing a set of blocks, how can we impl `StateProvider` for the in memory rep. of a few blocks?

Reply:
we're still lacking this feature: https://github.com/paradigmxyz/reth/issues/9614
to unblock this pr we can ignore the memory state for now and track in an issue instead

Reply:
ok, I've left todo's about this